fineract-provider/src/main/java/org/apache/fineract/accounting/producttoaccountmapping/domain/ProductToGLAccountMappingRepository.java

```diff
@@ -30,8 +30,9 @@
     ProductToGLAccountMapping findByProductIdAndProductTypeAndFinancialAccountTypeAndPaymentTypeId(Long productId, int productType,
             int financialAccountType, Long paymentType);
 
-    ProductToGLAccountMapping findByProductIdAndProductTypeAndFinancialAccountTypeAndChargeId(Long productId, int productType,
-            int financialAccountType, Long chargeId);
+    @Query("select mapping from ProductToGLAccountMapping mapping where mapping.productId= :productId and mapping.productType= :productType and mapping.financialAccountType= :financialAccountType and mapping.charge.id= :chargeId")
+    ProductToGLAccountMapping findProductIdAndProductTypeAndFinancialAccountTypeAndChargeId(@Param("productId") Long productId, @Param("productType") int productType,
+            @Param("financialAccountType") int financialAccountType, @Param("chargeId") Long ChargeId);
```

Comment:
@maektwain I'd like to better understand what problem this fixes how... At first glance, to me this looks like we are just writing an explicit `@Query` for what Spring Data otherwise automagically internally generates with from its `findBy` syntax... can you help me understand what I'm missing? What Query does the current version generate, and what is wrong with it which this explicit version fixes? How is the auto-generated query wrong?

Reply:
So, with findBy  query it does not return data , it return null value, even data is available, making it more dumb and simple.

Reply:
So, I tested it across steps to reproduce the above  now.

Create a savings product, define accounting and associate charges with advanced accounting mapping to Income and Fees.

Make a transaction, Charges associate to advanced accounting should go to those headers defined in previous step. 

Debugged and realised its coming null and with this query its appropriately returning and posting in journal entries.

Reply:
> So, with findBy query it does not return data , it return null value, even data is available, making it more dumb and simple.
 
I found this puzzling, and had a closer look... and noticed that `AndChargeId` is curious - there is no `chargeId` getter and field on `ProductToGLAccountMapping` - only a `charge`. Do you want to try to see if just renaming the method instead of having to manually specify the query also fixes this? If that is so, then I'm interested how to find a way to validate those Spring Data Repository interfaces...

>  So, I tested it across steps to reproduce the above now.

do you think you would be able to write an integration test for this problem in this PR? Just asking!

Reply:
I think no need , the test is already there I guess , if it's  not there then I can write.

Reply:
So, if we rename the method then we eventually have to change the existing method sigs, since it uses the ChargeId , and then in that case we need to pass the Charge. This fixes the problem.

Reply:
Fields Types are different and usage is also different if you look here 

File is AccountingProcessorHelper.java

``` java
if (accountMappingTypeId == CashAccountsForLoan.INCOME_FROM_FEES.getValue()
                || accountMappingTypeId == CashAccountsForLoan.INCOME_FROM_PENALTIES.getValue()) {
            final ProductToGLAccountMapping chargeSpecificIncomeAccountMapping = this.accountMappingRepository
                    .findByProductIdAndProductTypeAndFinancialAccountTypeAndChargeId(loanProductId, PortfolioProductType.LOAN.getValue(),
                            accountMappingTypeId, chargeId);
            if (chargeSpecificIncomeAccountMapping != null) {
                accountMapping = chargeSpecificIncomeAccountMapping;
            }
        }
```
It uses the chargeId 

and perviously it gets passed here 

```
final GLAccount chargeSpecificAccount = getLinkedGLAccountForSavingsCharges(savingsProductId, accountTypeToBeCredited.getValue(),
                chargePaymentDTO.getChargeId());
```

Reply:
@maektwain thanks for your explanations, I now (kind of) see what you mean - I had previously missed the `Charge` vs. `Long chargeId` ... then I guess this fix makes sense. I'm still puzzled as to why Spring doesn't flag this, but let's pick that up separately in FINERACT-998 later. I'm going to LGTM this PR, but not merge it, to let others review it as well, and give you a chance to still add an IT for it, that would be nice, but I don't want to blocking merging this even if no IT.

Reply:
@maektwain small practical thing, normally the commit message itself should include the JIRA, which is missing here, as you just added it to the PR title afterwards.. perhaps you want to amend the commit message?

Reply:
Thanks for the understanding . ACK.  

It has been that way, I will try to add IT too soon.

Reply:
Sure